### PR TITLE
Add coverage for AWS SQS

### PR DESCRIPTION
**Pull request by mtp1376**
Also:
- Add `normalizePolicy` function and use it in `policiesAreSame` function
- Use strongly-typed IAM policies across the codebase

### CODE REVIEW

File: test/modules/aws-sqs-integration.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+          SET is_default = TRUE
+          WHERE region = '${region}';
+      `,
+      undefined,
+      true,
+      () => ({ username, password }),
+    ),
+  );
+
+  itDocs('installs the SQS module', install(modules));
```
Comment (yrobla) [inline test/modules/aws-sqs-integration.ts:79]: it misses more itDocs here, to actually output code examples into documentation

**Review by yrobla — CHANGES_REQUESTED** [review]

File: test/modules/aws-sqs-integration.ts (hunk 1 above)
Comment (mtp1376) [reply test/modules/aws-sqs-integration.ts:79]: Added. Thanks for pointing out.

File: src/modules/aws_sdk/index.ts [hunk 2]
```diff
@@ -26,6 +26,7 @@ export class AwsSdkModule extends ModuleBase {
   invokeS3: AwsSdkInvoker;
   invokeSecretManager: AwsSdkInvoker;
   invokeSns: AwsSdkInvoker;
+  invokeSts: AwsSdkInvoker;
```
Comment (dfellis) [inline src/modules/aws_sdk/index.ts:29]: Does this work? Why does it say `Sts` instead of `Sqs`?

File: src/modules/interfaces.ts [hunk 3]
```diff
@@ -504,8 +504,8 @@ export class RpcBase implements RpcInterface {
 }
 
 export class AwsSdkInvoker extends RpcBase {
-  preTransactionCheck = PreTransactionCheck.WAIT_FOR_LOCK;
-  postTransactionCheck = PostTransactionCheck.UNLOCK_ALWAYS;
+  preTransactionCheck = PreTransactionCheck.NO_CHECK;
+  postTransactionCheck = PostTransactionCheck.NO_CHECK;
```
Comment (dfellis) [inline src/modules/interfaces.ts:508]: Should this change still be here?

File: src/services/aws_macros.ts [hunk 4]
```diff
@@ -74,6 +76,8 @@ export class AWS {
   route53Client: Route53;
   s3Client: S3;
   snsClient: SNS;
+  stsClient: STS;
```
Comment (dfellis) [inline src/services/aws_macros.ts:79]: Wait, what is `sts` and why didn't we need to import it in the `package.json`?

Also, why are we not exposing the SQS client in the `aws_sdk` module?

File: test/modules/aws-acm-import-integration.ts [hunk 5]
```diff
@@ -148,7 +148,7 @@ describe('AwsAcm Import Integration Testing', () => {
   itDocs(
     'import a certificate in non-default region',
     query(`
-    SELECT * FROM certificate_import('${cert}', '${key}', '${region}', '{}');
+    SELECT * FROM certificate_import('${cert}', '${key}', 'us-east-1', '{}');
```
Comment (dfellis) [inline test/modules/aws-acm-import-integration.ts:151]: Why this change?

File: src/modules/aws_sdk/index.ts (hunk 2 above)
Comment (mtp1376) [reply src/modules/aws_sdk/index.ts:29]: It does. It's for the `INSERT` trigger I already added (to get the account id using `sts.getCallerIdentity` function.

File: src/modules/interfaces.ts (hunk 3 above)
Comment (mtp1376) [reply src/modules/interfaces.ts:508]: No. Per our discussion on Wednesday's standup meeting this will be rolled back.

File: src/services/aws_macros.ts (hunk 4 above)
Comment (mtp1376) [reply src/services/aws_macros.ts:79]: Good point. Will expose SQS as well.

File: test/modules/aws-acm-import-integration.ts (hunk 5 above)
Comment (mtp1376) [reply test/modules/aws-acm-import-integration.ts:151]: It was using the default region. It should be non-default region instead.

**Review by yrobla — APPROVED** [review]

File: src/modules/aws_sqs/entity/queue.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+  /**
+   * @public
+   * The length of time, in seconds, for which Amazon SQS retains a message.
+   * @see https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-sqs/interfaces/createqueuecommandinput.html
+   */
+  @Min(60)
+  @Max(1_209_600)
+  @Column({
+    default: 345_600,
```
Comment (dfellis) [inline src/modules/aws_sqs/entity/queue.ts:96]: Nit: comment on what 345,600 seconds means (4 days)

```suggestion
    default: 345_600, // 4 days
```

File: src/modules/aws_sqs/entity/queue.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  })
+  receiveMessageWaitTimeSeconds: number;
+
+  /**
+   * @public
+   * The length of time, in seconds, for which Amazon SQS retains a message.
+   * @see https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-sqs/interfaces/createqueuecommandinput.html
+   */
+  @Min(60)
+  @Max(1_209_600)
```
Comment (dfellis) [inline src/modules/aws_sqs/entity/queue.ts:94]: Nit: comment on what 1,209,600 seconds means (14 days)

```suggestion
  @Max(1_209_600) // 14 days
```

Comment (dfellis) [reply src/modules/aws_sqs/entity/queue.ts:94]: (Also I am reminded of "A Fire Upon The Deep"'s use of Msec, Gsec, etc units once humanity no longer even remembers its home world but Unix somehow survives ;) )

File: src/modules/aws_sqs/sql/after_install.sql [hunk 8]
```diff
@@ -0,0 +1,21 @@
+CREATE FUNCTION
+  before_queue_insert_update () RETURNS TRIGGER LANGUAGE PLPGSQL AS $$
+BEGIN
+    -- can't create a non-fifo queue ending with .fifo and vice-versa
+    IF (NEW.name NOT LIKE '%.fifo') AND (NEW.fifo_queue) THEN
+        RAISE EXCEPTION 'The name of a fifo queue should end with .fifo';
+    END IF;
+    IF (NEW.name LIKE '%.fifo') AND (NOT NEW.fifo_queue) THEN
+        RAISE EXCEPTION 'Non-fifo queue name should not end with .fifo';
+    END IF;
```
Comment (dfellis) [inline src/modules/aws_sqs/sql/after_install.sql:10]: I *understand* why this is here, but wtf AWS. I am tempted to say "just auto add the `.fifo` prefix when needed and only ban having it explicitly in here" but that could cause confusion for other modules needing to reference this name so...

**Comment by mtp1376** [comment]
CodeBuild project tests are failing with this error:
```
[Container] 2023/02/17 13:08:53 Running command docker pull ubuntu:latest
39 | Error response from daemon: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
40
```

**Review by aguillenv — APPROVED** [review]

File: src/modules/aws_ecr/index.ts [hunk 9]
```diff
@@ -583,7 +580,7 @@ class RepositoryPolicyMapper extends MapperBase<RepositoryPolicy> {
         ctx,
         this.module.repository.generateId({ repositoryName: rp.repositoryName, region }),
       ));
-    out.policyText = rp?.policyText?.replace(/\n/g, '').replace(/\s+/g, ' ') ?? null;
+    out.policy = JSON.parse(rp?.policyText?.replace(/\n/g, '').replace(/\s+/g, ' ') ?? null);
```
Comment (dfellis) [inline src/modules/aws_ecr/index.ts:583]: Do we still need the `replace` calls here if you're `JSON.parse`ing it?

Also, on `JSON.parse` failure it `throw`s instead of returning something we'd replace with `null`, so you may want to move [`safeParse`](https://github.com/iasql/iasql/blob/main/src/modules/aws_acm/rpcs/common.ts#L1) into the `src/services` directory and use that instead?

File: src/modules/aws_sqs/mappers/queue.ts [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+    // https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_GetQueueAttributes.html
+    out.url = queueUrl;
+    out.name = parseArn(queueAttributes.QueueArn).resource;
+    out.fifoQueue = out.name.endsWith('.fifo');
+    out.visibilityTimeout = parseInt(queueAttributes.VisibilityTimeout, 10);
+    out.delaySeconds = parseInt(queueAttributes.DelaySeconds, 10);
+    out.receiveMessageWaitTimeSeconds = parseInt(queueAttributes.ReceiveMessageWaitTimeSeconds, 10);
+    out.messageRetentionPeriod = parseInt(queueAttributes.MessageRetentionPeriod, 10);
+    out.maximumMessageSize = parseInt(queueAttributes.MaximumMessageSize, 10);
+    out.policy = JSON.parse(queueAttributes.Policy);
```
Comment (dfellis) [inline src/modules/aws_sqs/mappers/queue.ts:68]: Same comment here about a `safeParse` being necessary.

File: src/modules/aws_vpc/entity/endpoint_gateway.ts [hunk 11]
```diff
@@ -54,8 +55,8 @@ export class EndpointGateway {
    * Complex type representing the policy associated to this gateway
    * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-resource-policies-examples.html
    */
-  @Column({ nullable: true })
-  policyDocument?: string;
+  @Column({ nullable: true, type: 'json' })
+  policyDocument?: Policy;
```
Comment (dfellis) [inline src/modules/aws_vpc/entity/endpoint_gateway.ts:59]: Do we want to normalize the naming across modules and make this just `policy`?

File: src/modules/aws_vpc/entity/endpoint_interface.ts [hunk 12]
```diff
@@ -76,8 +77,8 @@ export class EndpointInterface {
    * Complex type representing the policy associated to this endpoint
    * @see https://docs.aws.amazon.com/vpc/latest/privatelink/vpc-endpoints-access.html
    */
-  @Column({ nullable: true })
-  policyDocument?: string;
+  @Column({ nullable: true, type: 'json' })
+  policyDocument?: Policy;
```
Comment (dfellis) [inline src/modules/aws_vpc/entity/endpoint_interface.ts:81]: Same here. Just `policy`?

File: src/modules/aws_vpc/mappers/endpoint_gateway.ts [hunk 13]
```diff
@@ -53,7 +50,7 @@ export class EndpointGatewayMapper extends MapperBase<EndpointGateway> {
       (await this.module.vpc.db.read(ctx, this.module.vpc.generateId({ vpcId: eg.VpcId ?? '', region }))) ??
       (await this.module.vpc.cloud.read(ctx, this.module.vpc.generateId({ vpcId: eg.VpcId ?? '', region })));
     if (!out.vpc) return undefined;
-    out.policyDocument = eg.PolicyDocument;
+    out.policyDocument = eg.PolicyDocument ? JSON.parse(eg.PolicyDocument) : null;
```
Comment (dfellis) [inline src/modules/aws_vpc/mappers/endpoint_gateway.ts:53]: Double same: Just `policy` and using `safeParse`

File: src/modules/aws_vpc/mappers/endpoint_interface.ts [hunk 14]
```diff
@@ -121,7 +118,7 @@ export class EndpointInterfaceMapper extends MapperBase<EndpointInterface> {
       (await this.module.vpc.db.read(ctx, this.module.vpc.generateId({ vpcId: eg.VpcId ?? '', region }))) ??
       (await this.module.vpc.cloud.read(ctx, this.module.vpc.generateId({ vpcId: eg.VpcId ?? '', region })));
     if (!out.vpc) return undefined;
-    out.policyDocument = eg.PolicyDocument;
+    out.policyDocument = eg.PolicyDocument ? JSON.parse(eg.PolicyDocument) : null;
```
Comment (dfellis) [inline src/modules/aws_vpc/mappers/endpoint_interface.ts:121]: `policy`, `safeParse`...

File: test/modules/aws-codebuild-integration.ts [hunk 15]
```diff
@@ -350,8 +350,8 @@ phases:
   build:
     commands:
       - echo Building the Docker image...
-      - docker pull ubuntu:latest
-      - docker tag ubuntu:latest ' || (SELECT repository_uri FROM repository WHERE repository_name = '${dbAlias}' ) || ':latest
+      - docker pull public.ecr.aws/docker/library/ubuntu:latest
+      - docker tag public.ecr.aws/docker/library/ubuntu:latest ' || (SELECT repository_uri FROM repository WHERE repository_name = '${dbAlias}' ) || ':latest
```
Comment (dfellis) [inline test/modules/aws-codebuild-integration.ts:354]: # :star_struck:

**Comment by dfellis** [comment]
Pretty consistent failure. Looks like a diffing bug https://github.com/iasql/iasql/actions/runs/4206072004/jobs/7301643249#step:9:48173

**Review by yrobla — APPROVED** [review]
